presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyExpressions.java

```diff
@@ -234,26 +253,63 @@ public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, N
                     .collect(toImmutableList());
             Expression combinedUncorrelatedPredicates = combinePredicates(node.getType(), uncorrelatedPredicates);
 
-            // Do not simplify top level conjuncts if it would result in top level disjuncts
-            // Conjuncts are easier to process when pushing down predicates.
-            if (context.isRootNode() && flippedNodeType == OR && !combinedUncorrelatedPredicates.equals(FALSE_LITERAL)) {
-                return combinePredicates(node.getType(), predicates);
-            }
-
             return combinePredicates(flippedNodeType, ImmutableList.<Expression>builder()
                     .addAll(commonPredicates)
                     .add(combinedUncorrelatedPredicates)
                     .build());
         }
 
-        private static List<List<Expression>> getSubPredicates(List<Expression> predicates)
+        private static List<List<Expression>> getSubPredicates(LogicalBinaryExpression expression)
         {
-            return predicates.stream()
+            return extractPredicates(expression.getType(), expression).stream()
                     .map(predicate -> predicate instanceof LogicalBinaryExpression ?
                             extractPredicates((LogicalBinaryExpression) predicate) : ImmutableList.of(predicate))
                     .collect(toImmutableList());
         }
 
+        /**
+         * Applies the boolean distributive property.
+         *
+         * For example:
+         * ( A & B ) | ( C & D ) => ( A | C ) & ( A | D ) & ( B | C ) & ( B | D )
+         *
+         * Returns the original expression if the expression is non-deterministic or if the distribution will
+         * create more sub-expressions than the original.
+         */
+        private static Expression distributeIfPossible(LogicalBinaryExpression expression)
+        {
+            if (!DeterminismEvaluator.isDeterministic(expression)) {
+                // Do not distribute boolean expressions if there are any non-deterministic elements
+                // This can be optimized further, but let's not deal with it right now.
+                return expression;
+            }
+            List<Set<Expression>> subPredicates = getSubPredicates(expression).stream()
+                    .map(ImmutableSet::copyOf)
+                    .collect(toList());
+
+            int originalSubexpressions = subPredicates.stream()
+                    .mapToInt(Set::size)
+                    .reduce(Math::addExact)
+                    .getAsInt();
+            int newSubexpressions = subPredicates.stream()
+                    .mapToInt(Set::size)
+                    .reduce(Math::multiplyExact)
```

Comment:
The number of elementary subpredicates (e.g: `A, B, C, D`) will be larger. You have to multiply additionally by: `subPredicates.size()`.

For instance, for input expression:
`( A & B ) | ( C & D )` this will return 4
while it should be 8:
`( A | C ) & ( A | D ) & ( B | C ) & ( B | D )`